sqlmesh/core/engine_adapter/base.py

```diff
@@ -182,10 +182,18 @@ def _get_source_queries(
         batch_size = self.DEFAULT_BATCH_SIZE if batch_size is None else batch_size
         if isinstance(query_or_df, (exp.Query, exp.DerivedTable)):
             return [SourceQuery(query_factory=lambda: query_or_df)]  # type: ignore
+
         if not columns_to_types:
             raise SQLMeshError(
-                "It is expected that if a DF is passed in then columns_to_types is set"
+                "It is expected that if a DataFrame is passed in then columns_to_types is set"
+            )
+        if query_or_df.empty:
+            raise SQLMeshError(
+                "Cannot construct source query from an empty DataFrame. This error is commonly "
```

Comment:
would it be easier if we just handled the df empty case?

Reply:
discussed about it with Iaroslav too, but we concluded that this approach would be simpler as we don't have to think about edge cases.

Reply:
I _think_ it'd involve tracking down uses of `_get_source_queries_and_columns_to_types` and handling the empty `SourceQuery` list case. E.g. in `EngineAdapter.ctas` we could create a table with the `CREATE TABLE <name> <schema>` syntax instead of a CTAS. It's just a bit of a PITA because there are quite a few places where `_get_source_queries_and_columns_to_types` is used.